File: python and java/岛屿的个数.java
public class Solution {
    /**
     * @param grid a boolean 2D matrix
     * @return an integer
     */
    public int numIslands(boolean[][] grid) {
        // Write your code here
        int r = grid.length;
        int count = 0;
        int l;
        if (r == 0) {
        	return count;
        }
        l = grid[0].length;
        for (int i = 0; i < r; i++) {
        	for (int j = 0; j < l; j++) {
        		if (grid[i][j] == true) {
        			count++;
        			grid[i][j] = false;
        			DFS(grid, i, j, r, l);
        		}
        	}
        }
        return count;
    }

    public void DFS(boolean[][] grid,int i, int j, int r, int l){
    	grid[i][j] = false;
    	if (i + 1 < r && grid[i + 1][j] == true) {
    		DFS(grid, i + 1, j, r, l);
    	}
    	if (j + 1 < l && grid[i][j + 1] == true) {
    		DFS(grid, i, j + 1, r, l);
    	}
    	if (i - 1 >= 0 && grid[i - 1][j] == true) {
    		DFS(grid, i - 1, j, r, l);
    	}
    	if (j - 1 >= 0 && grid[i][j - 1] == true) {
    		DFS(grid, i, j - 1, r, l);
    	}
    }
}

File: python and java/岛屿的个数.py
class Solution:
    # @param {boolean[][]} grid a boolean 2D matrix
    # @return {int} an integer
    def numIslands(self, grid):
        # Write your code here
        count = 0
        r = len(grid)
        if r == 0:
        	return count
        l = len(grid[0])
        for i in range(r):
        	for j in range(l):
        		if grid[i][j] == 1:
        			count = count + 1
        			grid[i][j] = 0
        			self.DFS(grid,i,j,r,l)
        return count

    def DFS(self,grid,i,j,r,l):
    	grid[i][j] = 0
    	if i + 1 < r and grid[i + 1][j] == 1:
    		self.DFS(grid,i + 1,j,r,l)
    	if j + 1 < l and grid[i][j + 1] == 1:
    		self.DFS(grid,i,j + 1,r,l)
    	if i - 1 >= 0 and grid[i - 1][j] == 1:
    		self.DFS(grid,i - 1,j,r,l)
    	if j - 1 >= 0 and grid[i][j - 1] == 1:
    		self.DFS(grid, i, j - 1, r, l)